### PR TITLE
Introduce `StructuredMatrixStyle{Matrix}` (fixes #33397)

### LINKED ISSUE #33397: Broadcast with LowerTriangular, UpperTriangular, and Diagonal gives Diagonal
Broadcasting addition for `UpperTriangular`, `LowerTriangular`, and `Diagonal` produces a `Diagonal`. Note that normal addition is fine, and if either of these matrices are excluded, the result is fine.

```julia
julia> using Random, LinearAlgebra
julia> Random.seed!(42);
julia> U = UpperTriangular(randn(3,3));
julia> L = LowerTriangular(randn(3,3));
julia> D = Diagonal(randn(3));
julia> U+L+D
3×3 Array{Float64,2}:
 0.857936  -0.299484  -0.468606
 1.08238    2.41289    0.156143
 0.187028   0.367563  -5.28356 
julia> U.+L.+D
3×3 Diagonal{Float64,Array{Float64,1}}:
 0.857936   ⋅         ⋅     
  ⋅        2.41289    ⋅     
  ⋅         ⋅       -5.28356
```

Version info:
```julia
julia> versioninfo()
Julia Version 1.3.0-rc2.0
Commit a04936e3e0 (2019-09-12 19:49 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.6.0)
  CPU: Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, haswell)
```
Also confirmed on 1.2.

### PR DESCRIPTION
**Pull request by goggle**
As discussed in #33397, I have added a new structured matrix style type `StructuredMatrixStyle{Matrix}` to solve the issue.

Unfortunately, this introduced a new bug:
```julia
julia> using LinearAlgebra

julia> D = Diagonal(rand(2));

julia> M = rand(2,2);

julia> map(+, D, M, D)
2×2 Array{Float64,2}:
 1.7611    0.190262
 0.611437  1.32978
```
The `map` method suddenly returned a `Diagonal` matrix. That's the reason why I had to exclude the new type from the `fzero(S::StructuredMatrix)` method.

### CODE REVIEW

**Comment by goggle** [comment]
Ok, so this change currently breaks some tests, I will investigate it later.

I'm looking forward to hear some opinions/suggestions, maybe @mbauman?

File: stdlib/LinearAlgebra/src/structuredbroadcast.jl [hunk 1]
```diff
@@ -8,7 +8,7 @@ struct StructuredMatrixStyle{T} <: Broadcast.AbstractArrayStyle{2} end
 StructuredMatrixStyle{T}(::Val{2}) where {T} = StructuredMatrixStyle{T}()
 StructuredMatrixStyle{T}(::Val{N}) where {T,N} = Broadcast.DefaultArrayStyle{N}()
 
-const StructuredMatrix = Union{Diagonal,Bidiagonal,SymTridiagonal,Tridiagonal,LowerTriangular,UnitLowerTriangular,UpperTriangular,UnitUpperTriangular}
+const StructuredMatrix = Union{Matrix,Diagonal,Bidiagonal,SymTridiagonal,Tridiagonal,LowerTriangular,UnitLowerTriangular,UpperTriangular,UnitUpperTriangular}
```
Comment (mbauman) [inline stdlib/LinearAlgebra/src/structuredbroadcast.jl]: I don't think we want this change — and I think if we remove this change, then we can also remove the change to `fzero`... and I think then tests will pass again.

**Comment by mbauman** [comment]
Nice.  Now we just need to make sure that the `StructuredMatrix{<:Matrix}` is "poisoning" and doesn't ever end up falling through to give us back the DefaultArrayStyle from https://github.com/JuliaLang/julia/pull/33506/files#diff-575e6d191af9976b0c13526975e893d4R49.

Here's a still-failing test case:

```Julia
julia> LowerTriangular(rand(3,3)) .+ UpperTriangular(rand(3,3)) .+ LowerTriangular(rand(3,3)) .+ LowerTriangular(rand(3,3))
3×3 LowerTriangular{Float64,Array{Float64,2}}:
 2.33006    ⋅         ⋅
 0.787058  1.09198    ⋅
 0.831832  0.674229  1.4079
```

I think we need:
```Julia
Broadcast.BroadcastStyle(T::StructuredMatrixStyle{<:Matrix}, ::StructuredMatrixStyle) = T
Broadcast.BroadcastStyle(::StructuredMatrixStyle, T::StructuredMatrixStyle{<:Matrix}) = T
Broadcast.BroadcastStyle(T::StructuredMatrixStyle{<:Matrix}, ::StructuredMatrixStyle{<:Matrix}) = T
```

**Comment by goggle** [comment]
Thank you @mbauman for this great guidance! Now everything seems to work fine.

I don't really know anything about the backport policy of Julia. But since this fixes a bug which silently produced a wrong result, it might be a candidate for it?

**Comment by StefanKarpinski** [comment]
A bit complex for a pure bugfix, but may be internal enough. Have marked for triage to discuss.

**Comment by mbauman** [comment]
Let's let it sit on master for a little while before considering a back port.

Thanks @goggle!

**Comment by sdaxen** [comment]
Thanks everyone and especially @goggle for getting this fixed!!

**Pull request merged**